Integrate fuzzers from `google/oss-fuzz` repository + improve those fuzzers on top (#209)
uriparser/uriparser: CMakeLists.txt

option(URIPARSER_SHARED_LIBS "Build shared libraries (rather than static ones)" ${_URIPARSER_SHARED_LIBS_DEFAULT})
option(URIPARSER_BUILD_DOCS "Build API documentation (requires Doxygen, Graphviz, and (optional) Qt's qhelpgenerator)" ON)
option(URIPARSER_BUILD_TESTS "Build test suite (requires GTest >=1.8.0)" ON)
option(URIPARSER_BUILD_FUZZERS "Build fuzzers (requires Clang)" OFF)

> option(URIPARSER_OSSFUZZ_BUILD "Build fuzzers via ossfuzz" OFF)

Reviewer: ```suggestion
option(URIPARSER_OSSFUZZ_BUILD "Build fuzzers via OSS-Fuzz" OFF)
```

Author: Corrected

    add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND})
endif()

#
# Fuzzers
#

> if(URIPARSER_BUILD_FUZZERS AND DEFINED ENV{LIB_FUZZING_ENGINE})

Reviewer: I would rather not want to error out without `LIB_FUZZING_ENGINE` but to link with `-fsanitize=fuzzer` instead for users outside of OSS-Fuzz. For Expat, there is a dedicated flag `EXPAT_OSSFUZZ_BUILD` that allows users tselect from those two flavors of fuzzing:

- https://github.com/libexpat/libexpat/blob/c1c1677657944cf15faf59374c5b67f0541246f9/expat/CMakeLists.txt#L723-L729
- https://github.com/libexpat/libexpat/blob/c1c1677657944cf15faf59374c5b67f0541246f9/expat/CMakeLists.txt#L735-L737
- https://github.com/libexpat/libexpat/blob/c1c1677657944cf15faf59374c5b67f0541246f9/expat/CMakeLists.txt#L746-L755

How do you feel about a similar approach with a second flag `URIPARSER_OSSFUZZ_BUILD` in general?

Author: I'm OK with this approach but I want to ask one question before. This approach requires a separate target for the library that will be linked with the fuzzer:

https://github.com/libexpat/libexpat/blob/c1c1677657944cf15faf59374c5b67f0541246f9/expat/CMakeLists.txt#L734

that duplicates the "regular" target with a couple of changes like `target_compile_options(fuzzpat PRIVATE -fsanitize=fuzzer-no-link)`. It means that `uriparser` library will be compiled twice if `URIPARSER_BUILD_FUZZERS=ON`. Are you OK with this?

Reviewer: @tyler92 thanks for asking! If our options are…

- a) add `-fsanitize=fuzzer-no-link` to mainline uriparser (and putting it into non-production state) or
- b) have two copies

…I'm for two copies. Are there alternatives? We can call the new target `fuzzparser` if you like :smile:

Author: OK, got it, agree with you. Thank you for the hint :+1:  I will make all necessary changes tomorrow

Reviewer: @tyler92 PS: I should note that de-facto code…

```cmake
            if(NOT EXPAT_OSSFUZZ_BUILD)
                target_compile_options(${target_name} PRIVATE -fsanitize=fuzzer-no-link)
                set_target_properties(${target_name} PROPERTIES LINK_FLAGS -fsanitize=fuzzer)
            endif()
```

…as a subset of https://github.com/libexpat/libexpat/blob/c1c1677657944cf15faf59374c5b67f0541246f9/expat/CMakeLists.txt#L746-L755 started puzzling me as self-conflicting and unintended. I might be missing something, I have a feeling that `fuzzer-no-link` should be dropped here because it's the user facing fuzzer binaries this is for. Does it seem right as is to you?

Author: I've added `URIPARSER_OSSFUZZ_BUILD` with the same logic as for Expat.

> I have a feeling that fuzzer-no-link should be dropped here because it's the user facing fuzzer binaries this is for. Does it seem right as is to you?

I think it's OK: we compile `.c`/`.cpp` files with `-fsanitize=fuzzer-no-link` (even the main fuzz `cpp` file that contains `LLVMFuzzerTestOneInput`) and link with `-fsanitize=fuzzer`

Reviewer: @tyler92 I don't get it — why would we do that? I would expect them to ideally match.

Author: @hartwork I'm not sure I understood your question, could you please clarify it?

Reviewer: @tyler92 the mix of `-fsanitize=fuzzer-no-link` and `-fsanitize=fuzzer` for the fuzzing CLI extecutables does not yet make sense to me. Why is it not 2x `-fsanitize=fuzzer`? https://llvm.org/docs/LibFuzzer.html#id24 would imply that for our case. I guess both work but one is better fitting our semantics, or I missing something?
I just want to understand it and then potentially copy the new version over to Expat.

Reviewer: For anyone else reading: we changed it to `-fsanitize=fuzzer` in the meantime.

    function (add_ansi_and_unicode_fuzzer name source)
        if(URIPARSER_BUILD_CHAR)
            add_executable(${name}_fuzzer ${source})
            target_link_libraries(${name}_fuzzer PRIVATE uriparser $ENV{LIB_FUZZING_ENGINE})
            target_compile_definitions(${name}_fuzzer PRIVATE URI_NO_UNICODE)
        endif()

        if(URIPARSER_BUILD_WCHAR_T)
            add_executable(${name}w_fuzzer ${source})
            target_link_libraries(${name}w_fuzzer PRIVATE uriparser $ENV{LIB_FUZZING_ENGINE})
            target_compile_definitions(${name}w_fuzzer PRIVATE URI_NO_ANSI)
        endif()
    endfunction()

    add_ansi_and_unicode_fuzzer(uri_free fuzz/FreeFuzzer.cpp)
    add_ansi_and_unicode_fuzzer(uri_parse fuzz/ParseFuzzer.cpp)

>     add_ansi_and_unicode_fuzzer(uri_dissect_query_malloc fuzz/DissectQueryMallocFuzzer.cpp)

Reviewer: Alphabetical order please :pray:

Author: Fixed

    add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND})
endif()

#
# Fuzzers
#
if(URIPARSER_BUILD_FUZZERS)
    if(NOT "${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
        message(FATAL_ERROR
            "Building fuzz targets without Clang (but ${CMAKE_C_COMPILER_ID}) is not "
            "supported. Please set -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++.")
    endif()

    if(URIPARSER_OSSFUZZ_BUILD AND NOT DEFINED ENV{LIB_FUZZING_ENGINE})
        message(FATAL_ERROR
            "OSS-Fuzz builds require the environment variable "
            "LIB_FUZZING_ENGINE to be set. If you are seeing this "

>             "warning, it points to a deeper problem in the ossfuzz "

Reviewer: ```suggestion
            "warning, it points to a deeper problem in the OSS-Fuzz "
```
(I should fix this in Expat, too.)

Author: Corrected

            "build setup.")
    endif()

    add_library(fuzzparser STATIC ${API_HEADER_FILES} ${LIBRARY_CODE_FILES})
    target_compile_definitions(fuzzparser PUBLIC URI_STATIC_BUILD)
    target_include_directories(fuzzparser
        PUBLIC
            ${CMAKE_CURRENT_SOURCE_DIR}/include
        PRIVATE
            ${CMAKE_CURRENT_BINARY_DIR}  # for UriConfig.h
    )

    if(NOT URIPARSER_OSSFUZZ_BUILD)
        target_compile_options(fuzzparser PRIVATE -fsanitize=fuzzer-no-link)
    endif()

>     function (add_fuzzer target_name source compile_definitions)

Reviewer: ```suggestion
    function(add_fuzzer target_name source compile_definitions)
```

Author: Corrected

        add_executable(${target_name} ${source})
        target_link_libraries(${target_name} PRIVATE fuzzparser)
        target_compile_definitions(${target_name} PRIVATE ${compile_definitions})
        if(URIPARSER_OSSFUZZ_BUILD)
            set_target_properties(${target_name} PROPERTIES LINK_FLAGS $ENV{LIB_FUZZING_ENGINE})
            set_target_properties(${target_name} PROPERTIES LINKER_LANGUAGE "CXX")
        else()
            target_compile_options(${target_name} PRIVATE -fsanitize=fuzzer-no-link)
            set_target_properties(${target_name} PROPERTIES LINK_FLAGS -fsanitize=fuzzer)

>         endif()
>     endfunction()

Reviewer: ```suggestion
        endif()
        set_property(TARGET ${target_name} PROPERTY RUNTIME_OUTPUT_DIRECTORY fuzz)        
    endfunction()
```
(It's a detail, but would be nice. Matches Expat.)

Author: Added

>     function (add_ansi_and_unicode_fuzzer name source)

Reviewer: ```suggestion
    function(add_ansi_and_unicode_fuzzer name source)
```

Author: Corrected

        if(URIPARSER_BUILD_CHAR)
            add_fuzzer(${name}_fuzzer ${source} URI_NO_UNICODE)
        endif()

        if(URIPARSER_BUILD_WCHAR_T)
            add_fuzzer(${name}w_fuzzer ${source} URI_NO_ANSI)
        endif()
    endfunction()

    add_ansi_and_unicode_fuzzer(uri_dissect_query_malloc fuzz/DissectQueryMallocFuzzer.cpp)
    add_ansi_and_unicode_fuzzer(uri_free fuzz/FreeFuzzer.cpp)
    add_ansi_and_unicode_fuzzer(uri_parse fuzz/ParseFuzzer.cpp)
else()
    if(URIPARSER_OSSFUZZ_BUILD)
        message(SEND_ERROR

>                 "Attempting to perform an ossfuzz build without turning on the fuzzer build. "

Reviewer: ```suggestion
                "Attempting to perform an OSS-Fuzz build without turning on the fuzzer build. "
```

Author: Corrected

                "This is likely in error - consider adding "

>                 "-DURIPARSER_BUILD_FUZZERS=ON to your cmake execution.")

Reviewer: ```suggestion
                "-DURIPARSER_BUILD_FUZZERS=ON to your CMake execution.")
```

Author: Corrected

            "OSS-Fuzz builds require the environment variable "
            "LIB_FUZZING_ENGINE to be set. If you are seeing this "
            "warning, it points to a deeper problem in the OSS-Fuzz "
            "build setup.")
    endif()

    add_library(fuzzparser STATIC ${API_HEADER_FILES} ${LIBRARY_CODE_FILES})
    target_compile_definitions(fuzzparser PUBLIC URI_STATIC_BUILD)
    target_include_directories(fuzzparser
        PUBLIC
            ${CMAKE_CURRENT_SOURCE_DIR}/include
        PRIVATE
            ${CMAKE_CURRENT_BINARY_DIR}  # for UriConfig.h
    )

    if(NOT URIPARSER_OSSFUZZ_BUILD)
        target_compile_options(fuzzparser PRIVATE -fsanitize=fuzzer-no-link)
    endif()

    function(add_fuzzer target_name source compile_definitions)
        add_executable(${target_name} ${source})
        target_link_libraries(${target_name} PRIVATE fuzzparser)
        target_compile_definitions(${target_name} PRIVATE ${compile_definitions})
        if(URIPARSER_OSSFUZZ_BUILD)
            set_target_properties(${target_name} PROPERTIES LINK_FLAGS $ENV{LIB_FUZZING_ENGINE})
            set_target_properties(${target_name} PROPERTIES LINKER_LANGUAGE "CXX")
        else()

>             target_compile_options(${target_name} PRIVATE -fsanitize=fuzzer-no-link)

Reviewer: ```suggestion
            target_compile_options(${target_name} PRIVATE -fsanitize=fuzzer)
```

?

Related to https://github.com/uriparser/uriparser/pull/209#discussion_r1903331297

Author: I checked it - you are right, it works well. Applied your suggestion

message(STATUS "    Code for wchar_t * ... ${URIPARSER_BUILD_WCHAR_T}")
message(STATUS "    Tools ................ ${URIPARSER_BUILD_TOOLS}")
message(STATUS "    Test suite ........... ${URIPARSER_BUILD_TESTS}")

> message(STATUS "    Fuzzing tests ........ ${URIPARSER_BUILD_FUZZERS}")

Reviewer: ```suggestion
message(STATUS "    Fuzzers .............. ${URIPARSER_BUILD_FUZZERS}")
```

(Because these are not tests to me.)

Author: Fixed
